iree/turbine/kernel/wave/templates/quantized_attention.py

```diff
@@ -0,0 +1,225 @@
+from iree.turbine.kernel.wave.utils.mma_utils import (
+    get_mfma_load_elems_per_thread,
+    get_mfma_store_elems_per_thread,
+)
+from .attention_common import AttentionShape
+import math
+from iree.turbine.kernel.wave.utils.general_utils import torch_dtype_to_wave
+
+
+def get_brevitas_pertensor_fp8_attention_kernel(
+    shape: AttentionShape,
+    mfma_variant: MMAType,
+    logit_dtype: torch.dtype = torch.float16,
+    f8_dtpye: torch.dtype = torch.float8_e4m3fnuz,
+    dynamic_dims: bool = False,
+    is_causal: bool = False,
+    q_scale=1.0,
+    k_scale=1.0,
+    v_scale=1.0,
+):
+    # Input sizes
+    B = tkl.sym.B
+    M = tkl.sym.M
+    N = tkl.sym.N
+    K1 = tkl.sym.K1
+    K2 = tkl.sym.K2
+    # Workgroup tile sizes
+    BLOCK_B = tkl.sym.BLOCK_B
+    BLOCK_M = tkl.sym.BLOCK_M
+    BLOCK_N = tkl.sym.BLOCK_N
+    BLOCK_K2 = tkl.sym.BLOCK_K2
+    # Address space (for GPU, shared(1) or global(0))
+    ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+    # Other hyperparameters
+    LOAD_ELEMS_PER_THREAD_QK = index_symbol("LOAD_ELEMS_PER_THREAD_QK")
+    LOAD_ELEMS_PER_THREAD_PV = index_symbol("LOAD_ELEMS_PER_THREAD_PV")
+    STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+    # Expose user-constraints
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.WorkgroupConstraint(B, BLOCK_B, 2)]
+    constraints += [tkw.TilingConstraint(K2, BLOCK_K2)]
+    constraints += [tkw.WaveConstraint(M, BLOCK_M / 4)]
+    constraints += [tkw.WaveConstraint(N, BLOCK_N / 1)]
+
+    if mfma_variant[0] == MMAType.F32_16x16x32_F8:
+        Mvec = 16
+        Nvec = 16
+    if mfma_variant[0] == MMAType.F32_32x32x16_F8:
+        Mvec = 32
+        Nvec = 32
+
+    constraints += [
+        tkw.HardwareConstraint(
+            threads_per_wave=64,
+            waves_per_block=(4, 1, 1),
+            mma_type=mfma_variant[1],
+            vector_shapes={B: 0, M: Mvec, N: Nvec},
+        )
+    ]
+
+    if dynamic_dims:
+        constraints += [tkw.Assumption(K2 > BLOCK_K2 * 4)]
+
+    i = tkw.IndexMapping.iterator(0)
+    j = tkw.IndexMapping.iterator(1)
+    k = tkw.IndexMapping.iterator(2)
+    mapping = tkw.IndexMapping(
+        num_iterators=3, inputs={B: i, N: j, M: k}, outputs={B: i, M: k, N: j}
+    )
+
+    # Value tensor mapping to transpose for efficient computation if the input is
+    # not already transposed.
+    v_mapping = tkw.IndexMapping(
+        num_iterators=3, inputs={B: i, N: j, K2: k}, outputs={B: i, N: j, K2: k}
+    )
+
+    # Setting up input DTYPE
+    LOGIT_DTYPE = torch_dtype_to_wave(logit_dtype)
+
+    # Setting up FP8 scaling
+    LOG2E = 1.44269504089
+    DK_SQRT = math.sqrt(1.0 / shape.head_size)
+    F8_DTYPE = torch_dtype_to_wave(f8_dtpye)
+    F8_MAX = torch.finfo(f8_dtpye).max
+
+    # maximum expected value from attention softmax
+    ATTENTION_SOFTMAX_MAX = 1.0
+
+    # FP8 offset
```

Comment:
Doesnt have to be in this PR, but do you think you could add a doc on FP8 Attention quantization? I think this would be very interesting for folks.

Reply:
will add in separate PR!